File: Module 4/46_Graphs/Python/Graphs.py
class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adjMatrix = [[0 for _ in range(vertices)] for _ in range(vertices)]

    def add_edge(self, src, des):
        # Undirected Graph
        self.adjMatrix[src][des] = 1
        self.adjMatrix[des][src] = 1

    def print_graph(self):
        for row in self.adjMatrix:
            print(" ".join(map(str, row)))

    def DFS(self, vertex, visited):
        visited[vertex] = True
        print(vertex, end=" ")

        for i in range(self.vertices):
            if self.adjMatrix[vertex][i] == 1 and not visited[i]:
                self.DFS(i, visited)

    def BFS(self, vertex):
        visited = [False] * self.vertices
        queue = []
        visited[vertex] = True
        queue.append(vertex)

        while queue:
            curr = queue.pop(0)
            print(curr, end=" ")

            for i in range(self.vertices):
                if self.adjMatrix[curr][i] == 1 and not visited[i]:
                    visited[i] = True
                    queue.append(i)


if __name__ == "__main__":
    g = Graph(5)

    g.add_edge(0, 1)
    g.add_edge(0, 4)
    g.add_edge(1, 3)
    g.add_edge(1, 4)
    g.add_edge(2, 3)
    g.add_edge(3, 4)

    print("Adjacency Matrix Representation of the Graph")
    g.print_graph()

    print("\nDepth First Search: ", end="")
    visited = [False] * 5
    g.DFS(0, visited)

    print("\nBreadth First Search: ", end="")
    g.BFS(0)

File: Module 4/46_Graphs/Cpp/Graphs.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Graph {
    int vertices;
    vector<vector<int>> adjMatrix;

public:
    // Constructor
    Graph(int v) {
        vertices = v;
        adjMatrix = vector<vector<int>>(v, vector<int>(v, 0));
    }

    void addEdge(int src, int des) {
        // Undirected Graph
        adjMatrix[src][des] = 1;
        adjMatrix[des][src] = 1;
    }

    void printGraph() {
        for (int i = 0; i < vertices; i++) {
            for (int j = 0; j < vertices; j++) {
                cout << adjMatrix[i][j] << " ";
            }
            cout << endl;
        }
    }

    void DFS(int vertex, vector<bool>& visited) {
        visited[vertex] = true;
        cout << vertex << " ";

        for (int i = 0; i < vertices; i++) {
            if (adjMatrix[vertex][i] == 1 && !visited[i]) {
                DFS(i, visited);
            }
        }
    }

    void BFS(int vertex) {
        vector<bool> visited(vertices, false);
        queue<int> q;
        visited[vertex] = true;
        q.push(vertex);

        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            cout << curr << " ";

            for (int i = 0; i < vertices; i++) {
                if (adjMatrix[curr][i] == 1 && !visited[i]) {
                    visited[i] = true;
                    q.push(i);
                }
            }
        }
    }
};

int main() {
    Graph g(5);

    g.addEdge(0, 1);
    g.addEdge(0, 4);
    g.addEdge(1, 3);
    g.addEdge(1, 4);
    g.addEdge(2, 3);
    g.addEdge(3, 4);

    cout << "Adjacency Matrix Representation of the Graph" << endl;
    g.printGraph();

    cout << "\nDepth First Search: ";
    vector<bool> visited(5, false);
    g.DFS(0, visited);

    cout << "\nBreadth First Search: ";
    g.BFS(0);

    return 0;
}
